Chapter5/changeFreq.erl

```diff
@@ -0,0 +1,117 @@
+-module(changeFreq).
+-export([start/0, stop/0, allocate/0, deallocate/1]).
+-export([init/0]).
+
+%% These are the start functions used to create and
+%% initialize the server.
+
+start() ->
+	register(changeFreq, spawn(changeFreq, init, [])).
+
+init() ->
+	Frequencies = {get_frequencies(), []},
+	loop(Frequencies).
+
+% Hard Coded
+get_frequencies() -> [10,11,12,13,14,15].
+
+
+
+%% The client Functions
+stop() 			 -> call(stop).
+
+allocate() 		 -> call(allocate).
+
+deallocate(Freq) -> call({deallocate, Freq}).
+
+%% We hide all message passing and the message
+%% protocol in a functional interface.
+call(Message) ->
+	changeFreq ! {request, self(), Message},
+	receive
+		{reply, Reply} -> Reply
+	end.
+
+
+
+
+%% The Main Loop
+loop(Frequencies) ->
+	receive
+		{request, Pid, allocate} ->
+			{NewFrequencies, Reply} = allocate(Frequencies, Pid),
+			reply(Pid, Reply),
+			loop(NewFrequencies);
+
+		{request, Pid , {deallocate, Freq}} ->
+			{NewFrequencies, Reply} = deallocate(Frequencies, Freq),
+			reply(Pid, Reply),
+			loop(NewFrequencies);
+
+		{request, Pid, stop} ->
+			{_, Allocated} = Frequencies,
+			case Allocated == [] of
```

Comment:
try to avoid using `case` statements to match on boolean, erlang provides powerful pattern matching so that you can match on complex structures:
```erlang
case Frequencies of
    {_, [] = _Allocated} ->
        reply(Pid, ok);
    {_, _Allocated} ->
        reply(Pid, {error, frequencies_still_allocated}),
        loop(Frequencies)
end
```